### PR TITLE
:tada: Source MongoDB: adding level parameter to define max-depth

### PR DESCRIPTION
**Pull request by hash-data**
## What
Adding Max_Level so that the user can configure how much depth the user has to flatten the data for warehouses query.
For extensive data having nested documents the Mongo connector takes a lot of time and Memory.

Use the mongo cursor to find keys so that all collections will not be taken in RAM at once saving RAM.

Added step for also considering empty string key.
Screenshot 2023-02-28 at 12 22 25 PM
## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Comment by hash-data** [comment]
Hi @marcosmarxm please take a look here
I am having trouble with it.
Thanks

**Comment by hash-data** [comment]
Hi, @marcosmarxm  I have run the integration tests 
but the integration test also failing on current master branch too
can you check it.

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io.airbyte.integrations.source.mongodb/MongoDbSource.java [hunk 1]
```diff
@@ -102,9 +102,13 @@ protected List<TableInfo<CommonField<BsonType>>> discoverInternal(final MongoDat
     final List<TableInfo<CommonField<BsonType>>> tableInfos = new ArrayList<>();
 
     final Set<String> authorizedCollections = getAuthorizedCollections(database);
+    final var config = database.getSourceConfig();
+    // check if maxDepth Level exist if not then set default to 2
+    final var maxDepthLevel =  config.has(MongoUtils.MAX_DEPTH_LEVEL_READ) ? config.get(MongoUtils.MAX_DEPTH_LEVEL_READ).asInt(): 2;
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io.airbyte.integrations.source.mongodb/MongoDbSource.java:107]:
```suggestion
    final int DEFAULT_DEPTH_LEVEL_READ  = 2; 
    final var maxDepthLevel =  config.has(MongoUtils.MAX_DEPTH_LEVEL_READ) ? config.get(MongoUtils.MAX_DEPTH_LEVEL_READ).asInt(): DEFAULT_DEPTH_LEVEL_READ;
```

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
small comment and waiting JDBC team to validate the change.

**Comment by marcosmarxm** [comment]
@hash-data how can I make to work to flat all data without using the max-depth? I think this is the problem which will limit users with this feature.

**Comment by marcosmarxm** [comment]
@hash-data what happen if I use a max-depth but my data doesn't have the depth?

**Comment by hash-data** [comment]
> @hash-data what happen if I use a max-depth but my data doesn't have the depth?

there are two cases either subfield exists greater than max-depth var then it will go till max-depth and if max-depth is greater than the subfield existence then it will only go till the subfield existance.

**Comment by hash-data** [comment]
> @hash-data how can I make to work to flat all data without using the max-depth? I think this is the problem which will limit users with this feature.

this needs to be handled 
one solution is to define a large max depth and then subfield existence or the other is to take input of  boolean full_depth_scan

**Comment by marcosmarxm** [comment]
I'd like the implementation doesn't change current usage of users which isn't the case. I think the default should be go max depth by default.

**Comment by hash-data** [comment]
@marcosmarxm  and team check this PR

**Comment by jdpgrailsdev** [comment]
@hash-data Thanks for the PR!  The team is currently working on a new, improved MongoDB source that will no longer recursively search for fields as part of the discovery step.  After research, we have decided that this is the best approach going forward.  Our currently plan is to release the new connector by the end of September.  Thanks for your patience and contribution.

**Pull request closed without merging**